src/Text/Parsing/Parser.purs

```diff
@@ -120,10 +125,77 @@ instance monadPlusParserT :: Monad m => MonadPlus (ParserT s m)
 instance monadTransParserT :: MonadTrans (ParserT s) where
   lift = ParserT <<< lift <<< lift
 
+instance compactableParserT :: Monad m => Compactable (ParserT s m) where
+  compact p1 = do
+    state <- get
+    p1 >>= case _ of
+      Just r -> pure r
+      Nothing -> put state *> fail "Parse returned Nothing"
+  separate p1 =
+    { left: do
+      state <- get
+      p1 >>= case _ of
+        Left r -> pure r
+        Right r -> put state *> fail "Parse returned Right"
+    , right: do
+      state <- get
+      p1 >>= case _ of
+        Left r -> put state *> fail "Parse returned Left"
```

Comment:
The point in getting the state is that it's done before the parser `p1` is run. The way you changed it makes it so that the getting and setting happen right after eachother, which would be the same as just `fail`ing right away and not doing anything else. The reason i'm getting the state before running the parser is that when the parser fails, I believe the right move is to error at the position from *before* the parser. Let's say you parse "34" with `digit *> filter (_ == 1) digit` (`digit` is from Test.Main and just parses a single digit 1-9). It will error at position 2, as it should. (Reason being that the character `4` is what's wrong). If we do not get the state beforehand and set it when erroring, the error would report position 3, which would be incorrect. The reason for grabbing the whole state is that if the result is filtered away, I want the parser to act as if it wasn't run at all, except for erroring of course. It might be better to just get the position though, and then fail with that position? (not altering other parts of the state.) Then it would be possible to decide whether to have it backtrack or not when using it with `<|>`, simply by putting `try` in front of it or not.